Merging credential_response_encryption parameters into a single object (#136)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

* `proof`: OPTIONAL. Object containing the proof of possession of the cryptographic key material the issued Credential would be bound to.  The `proof` object MUST contain a following claim:
    * `proof_type`: REQUIRED. String denoting the key proof type. The value of this claim determines other claims in the key proof object and its respective processing rules. Key proof types defined in this specification can be found in (#proof_types).
* `credential_identifier`: REQUIRED when `credential_identifier` was returned from the Token Response. MUST NOT be used otherwise. String that identifies a Credential that is being requested to be issued. When this parameter is used, the `format` parameter and any other Credential format specific set of parameters such as those defined in (#format_profiles) MUST NOT be present.

> * `credential_response_encryption`: OPTIONAL. Object containing information for encrypting the Credential Response. If omitted, encryption is not performed.

Reviewer: ```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information for encrypting the Credential Response. If this request element is not present, the corresponding credential response returned is not encrypted.
```

> * `credential_response_encryption`: OPTIONAL. Object containing information for encrypting the Credential Response. If omitted, no encryption is intended to be performed.

Reviewer 2: ```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information for encrypting the Credential Response. If omitted, encryption is not performed.
```

    * `jwk`: REQUIRED. Object containing a single public key as a JWK used for encrypting the Credential Response.

>     * `alg`: OPTIONAL. JWE [@!RFC7516] `alg` algorithm [@!RFC7518] for encrypting Credential Responses. The default value is `xxx`.

Reviewer 3: ```suggestion
    * `jwk`: Object containing a single public key as a JWK used for encrypting the Credential Response.
    * `alg`: JWE [@!RFC7516] `alg` algorithm [@!RFC7518] for encrypting Credential Responses. The default value is `xxx`.
```

> * `credential_response_encryption`: OPTIONAL. Object containing information for encrypting the Credential Response. If omitted, encryption is not performed.
>     * `jwk`: REQUIRED. Object containing a single public key as a JWK used for encrypting the Credential Response.
>     * `alg`: REQUIRED. JWE [@!RFC7516] `alg` algorithm [@!RFC7518] for encrypting Credential Responses.
>     * `enc`: REQUIRED. JWE [@!RFC7516] `enc` algorithm [@!RFC7518] for encrypting Credential Responses.

Reviewer 3: ```suggestion
* `credential_response_encryption`: OPTIONAL. Object containing information for encrypting the Credential Response. If omitted, encryption is not performed.
    * `jwk`: Object containing a single public key as a JWK used for encrypting the Credential Response.
    * `alg`: JWE [@!RFC7516] `alg` algorithm [@!RFC7518] for encrypting Credential Responses.
    * `enc`: JWE [@!RFC7516] `enc` algorithm [@!RFC7518] for encrypting Credential Responses.
```

Reviewer 3: I don't think we will need the keywords here. Those are CONDITIONAL based on the presence of their parent.

Author: I disagree, currently all subclaims are specified as OPTIONAL or REQUIRED throughout the whole spec

Reviewer 3: @paulbastian @Sakurann Ok, then let's keep it the way it is written now. My intention is not to change the entire spec. I withdraw this comment.

Reviewer 4: On editors call we seemed to have a consensus around "REQUIRED when credential_response_encryption parameter is present."

>     * `enc`: OPTIONAL. JWE [@!RFC7516] `enc` algorithm [@!RFC7518] for encrypting Credential Responses. The default value is `A256GCM`.

Author: ```suggestion
    * `enc`: REQUIRED. JWE [@!RFC7516] `enc` algorithm [@!RFC7518] for encrypting Credential Responses.
```

>     * `enc`: OPTIONAL. JWE [@!RFC7516] `enc` algorithm [@!RFC7518] for encrypting Credential Responses. The default value is `A256GCM`.

Reviewer 3: ```suggestion
    * `enc`: JWE [@!RFC7516] `enc` algorithm [@!RFC7518] for encrypting Credential Responses. The default value is `A256GCM`.
```

>     * `enc`: OPTIONAL. JWE [@!RFC7516] `enc` algorithm [@!RFC7518] for encrypting Credential Responses. The default value is `A256GCM`.

Reviewer 5: I would go for having this REQUIRED as well, making default values according to specific algs, today, we may require errata in the future when something might became vulnerable or deprecable.

Reviewer 3: if we keep those key words in general, then we need to make all of them REQUIRED.

Author: This is already resolved in the PR, closing this conversation

In other cases, the Credential Issuer MAY NOT be able to immediately issue a requested Credential and would want to send a `transaction_id` parameter to the Client to be used later to receive a Credential when it is ready. The HTTP status code MUST be 202 (section 10.2.3 of [@!RFC2616]).

> If the Client requested encrypted responses, the Credential Issuer MUST encode the information in the Credential Response as a JWT using the `credential_response_encryption` parameter in the Credential Request. If the Credential Response is encrypted, the media type of the response MUST bet set to `application/jwt`. If encryption was negotiated in the Credential Request and the Credential Response is not encrypted, the Client SHOULD reject the Credential Response.

Reviewer 4: Reworded slightly:

- make it clear that credential_response_encryption is how the client requests encryption
- change 'negotiated' to 'requested' as I think the latter better reflects what happens

```suggestion
If the Client requested an encrypted response by including the `credential_response_encryption` object in the request, the Credential Issuer MUST encode the information in the Credential Response as a JWT using the  parameters from the `credential_response_encryption` object. If the Credential Response is encrypted, the media type of the response MUST bet set to `application/jwt`. If encryption was requested in the Credential Request and the Credential Response is not encrypted, the Client SHOULD reject the Credential Response.
```

> If the Client requested encrypted responses, the Credential Issuer MUST encode the information in the Credential Response as a JWT with `alg`, `enc` and using the public key specified by `jwk` from the `credential_response_encryption` object in the Credential Request. If the Credential Response is encrypted, the media type of the response MUST bet set to `application/jwt`. If encryption was negotiated in the Credential Request and the Credential Response is not encrypted, the Client SHOULD reject the Credential Response.

Reviewer 5: Does this mean that the the response is an encrypted json and not a Netsted JWT, signed and then encrypted?
I would give more clarification, as OIDC with the userinfo response (or id token) made:

> If the UserInfo Response is signed and/or encrypted, then the Claims are returned in a JWT and the content-type MUST be application/jwt. The response MAY be encrypted without also being signed. If both signing and encryption are requested, the response MUST be signed then encrypted, with the result being a Nested JWT, as defined in [[JWT]](https://openid.net/specs/openid-connect-core-1_0.html#JWT).

@sakimura @selfissued ^

Author: I think this is about encryption only, this is how I read it

Reviewer 2: @peppelinux i think here we are trying to talk about the case when response is only encrypted and not signed. and that probably can be made clearer.

> If the Client requested encrypted responses, the Credential Issuer MUST encode the information in the Credential Response as a JWT with `alg`, `enc` and using the public key specified by `jwk` from the `credential_response_encryption` object in the Credential Request. If the Credential Response is encrypted, the media type of the response MUST bet set to `application/jwt`. If encryption was negotiated in the Credential Request and the Credential Response is not encrypted, the Client SHOULD reject the Credential Response.

Reviewer 3: ```suggestion
If the Client requested encrypted responses, the Credential Issuer MUST encode the information in the Credential Response as a JWT using the `credential_response_encryption` parameter in the Credential Request. If the Credential Response is encrypted, the media type of the response MUST bet set to `application/jwt`. If encryption was negotiated in the Credential Request and the Credential Response is not encrypted, the Client SHOULD reject the Credential Response.
```
